lib/common/src/Dom/Document.php

```diff
@@ -0,0 +1,1326 @@
+     * @param string $element   Element of the tag.
+     * @param string $attribute Attribute that the tag contains.
+     * @return string[] The requested tags. Returns an empty array if none found.
+     */
+    private function findTags($content, $element, $attribute = null)
+    {
+        $matches = [];
+        $pattern = empty($attribute)
+            ? sprintf(
+                self::HTML_FIND_TAG_WITHOUT_ATTRIBUTE_PATTERN,
+                preg_quote($element, self::HTML_FIND_TAG_DELIMITER)
+            )
+            : sprintf(
+                self::HTML_FIND_TAG_WITH_ATTRIBUTE_PATTERN,
+                preg_quote($element, self::HTML_FIND_TAG_DELIMITER),
+                preg_quote($attribute, self::HTML_FIND_TAG_DELIMITER)
+            );
+
+        if (preg_match($pattern, $content, $matches)) {
+            return $matches;
+        }
+
+        return [];
+    }
+
+    /**
+     * Find a given tag with a given attribute.
+     *
+     * If multiple tags match, this method will only return the first one.
+     *
+     * @param string $content   Content in which to find the tag.
+     * @param string $element   Element of the tag.
+     * @param string $attribute Attribute that the tag contains.
+     * @return string|false The requested tag, or false if not found.
+     */
+    private function findTag($content, $element, $attribute = null)
+    {
+        $matches = $this->findTags($content, $element, $attribute);
+
+        if (empty($matches)) {
+            return false;
+        }
+
+        return $matches[0];
+    }
+
+    /**
+     * Extract an attribute value from an HTML tag.
+     *
+     * @param string $tag       Tag from which to extract the attribute.
+     * @param string $attribute Attribute of which to extract the value.
+     * @return string|false Extracted attribute value, false if not found.
+     */
+    private function extractValue($tag, $attribute)
+    {
+        $matches = [];
+        $pattern = sprintf(
+            self::HTML_EXTRACT_ATTRIBUTE_VALUE_PATTERN,
+            preg_quote($attribute, self::HTML_FIND_TAG_DELIMITER)
+        );
+
+        if (preg_match($pattern, $tag, $matches)) {
+            return empty($matches['full']) ? $matches['partial'] : $matches['full'];
+        }
+
+        return false;
+    }
+
+    /**
+     * Sanitize the encoding that was detected.
+     *
+     * If sanitization fails, it will return 'auto', letting the conversion
+     * logic try to figure it out itself.
+     *
+     * @param string $encoding Encoding to sanitize.
+     * @return string Sanitized encoding. Falls back to 'auto' on failure.
+     */
+    private function sanitizeEncoding($encoding)
+    {
+        if (! function_exists('mb_list_encodings')) {
+            return $encoding;
+        }
+
+        static $knownEncodings = null;
+
+        if (null === $knownEncodings) {
+            $knownEncodings = array_map('strtolower', mb_list_encodings());
```

Comment:
I'd prefer to have it `suggest`ed only and safeguard the code in case it is not available.

A hard requirement like this could fail unexpectedly when it would not actually need to, like using a CLI environment to build the plugin that doesn't have `mbstring` included, but wouldn't ever run the code anyway.

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/4019/commits/ed3416992bb0459192439fb151b4f0328d054807